### PR TITLE
🐭 Introduce connectors

### PR DESCRIPTION
**Pull request by yivlad**

### CODE REVIEW

File: .changeset/polite-wasps-worry.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+---
+'@usedapp/coinbase-connector': patch
+'@usedapp/portis-connector': patch
+'@usedapp/wallet-connect-connector': patch
+'@usedapp/core': patch
```
Comment (rzadp) [inline .changeset/polite-wasps-worry.md:5]: minor?

File: packages/connectors/coinbase-wallet/package.json [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  "main": "dist/src/index.js",
+  "types": "dist/src/index.d.ts",
+  "repository": "git@github.com:EthWorks/useDApp.git",
+  "author": "Ethworks",
+  "license": "MIT",
+  "publishConfig": {
+    "access": "public"
+  },
+  "dependencies": {
+    "@coinbase/wallet-sdk": "^3.3.0"
```
Comment (rzadp) [inline packages/connectors/coinbase-wallet/package.json:13]: Have you considered pinning exact version? We will have one less thing to consider when debugging bug reports.

File: packages/core/src/hooks/useBlockNumber.ts [hunk 3]
```diff
@@ -1,11 +1,61 @@
...
+        isMounted = false
+        unsub()
+      }
+    }
+
+    if (!connector) {
+      return
+    }
+    const unsub = connector.newBlock.on((blockNumber) => {
+      if (isMounted) {
```
[the remark this answers is not in the document]
Comment (yivlad) [reply packages/core/src/hooks/useBlockNumber.ts:49]: Exactly. It is definitely helping.

File: packages/core/src/providers/network/connectors/context.tsx [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        setLoading(false)
+      }
+    },
+    [setController, setLoading]
+  )
+
+  const activateBrowserWallet: ActivateBrowserWallet = useCallback(
+    async ({ type } = { type: 'metamask' }) => {
+      if (!connectors[type]) {
+        throw new Error(`Connector ${type} is not configured`)
```
[the remark this answers is not in the document]
Comment (yivlad) [reply packages/core/src/providers/network/connectors/context.tsx:81]: It is not - default config includes properly configured metamask connector

File: packages/core/src/providers/blockNumber/blockNumbers/provider.tsx [hunk 5]
```diff
@@ -16,11 +16,22 @@ export function BlockNumbersProvider({ children }: Props) {
...
 
   useEffect(() => {
+    let isMounted = true
     const onUnmount = Object.entries(networks).map(([chainId, provider]) =>
-      subscribeToNewBlock(provider, Number(chainId), dispatch, isActive)
+      subscribeToNewBlock(
+        provider,
+        Number(chainId),
+        (...args: Parameters<typeof dispatch>) => {
+          if (isMounted) {
```
[the remark this answers is not in the document]
Comment (yivlad) [reply packages/core/src/providers/blockNumber/blockNumbers/provider.tsx]: Yes

File: packages/connectors/portis/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    }
+  }
+
+  async activate(): Promise<void> {
+    try {
+      await this.init()
+      const accounts: string[] = await this.provider!.send('eth_accounts', [])
+      this.update.emit({ chainId: this.chainId, accounts })
+    } catch (e) {
+      console.log(e)
```
Comment (rzadp) [inline packages/connectors/portis/src/index.ts:41]: throw? report error? add a comment why catching?

File: packages/connectors/wallet-connect/src/index.ts [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+      const chainId: string = await this.provider!.send('eth_chainId', [])
+      const accounts: string[] = await this.provider!.send('eth_accounts', [])
+      this.update.emit({ chainId: parseInt(chainId), accounts })
+    } catch (e) {
+      console.debug(e)
+    }
+  }
+
+  async activate(): Promise<void> {
+    await this.init()
```
Comment (rzadp) [inline packages/connectors/wallet-connect/src/index.ts:38]: init not in try - catch? Not consistent with portis?

File: packages/core/src/providers/network/connectors/connectorController.ts [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+        if (chainId !== undefined) {
+          this.chainId = chainId
+        }
+        if (accounts !== undefined) {
+          this.accounts = accounts
+        }
+        this.emitUpdate()
+      },
+      () => {
+        if (this.connector instanceof MetamaskConnector && this._config.noMetamaskDeactivate) {
```
Comment (rzadp) [inline packages/core/src/providers/network/connectors/connectorController.ts:82]: is `this._config` already set at this point?

File: packages/core/src/providers/network/connectors/connectorController.ts [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+
+      return () => {
+        provider.off('block', listener)
+      }
+    })
+
+    this.emitUpdate()
+  }
+
+  async deactivate() {
```
Comment (rzadp) [inline packages/core/src/providers/network/connectors/connectorController.ts:108]: make sure to call it when users disconnects?

File: packages/core/src/providers/network/connectors/context.tsx [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+  deactivate: () => void
+  activateBrowserWallet: ActivateBrowserWallet
+  reportError: (error: Error) => void
+  isLoading: boolean
+}
+
+export const ConnectorContext = createContext<ConnectorContextValue>({
+  connector: undefined,
+  //eslint-disable-next-line @typescript-eslint/no-empty-function
+  activate: async () => {},
```
Comment (rzadp) [inline packages/core/src/providers/network/connectors/context.tsx:27]:
```suggestion
  activate: async () => void,
```
try this

File: packages/core/src/providers/network/connectors/context.tsx [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+          ? providerOrConnector
+          : new Web3Provider(providerOrConnector)
+        controller = new ConnectorController(new InjectedConnector(wrappedProvider))
+      }
+      setLoading(true)
+      try {
+        if (silently) {
+          await controller.activate((connector) => connector.connectEagerly())
+        }
+        {
```
Comment (rzadp) [inline packages/core/src/providers/network/connectors/context.tsx:64]: block needed?

File: packages/core/src/providers/network/connectors/implementations/metamask.ts [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+  }
+
+  const injectedProviders: any[] = (window?.ethereum as any).providers || []
+  const injectedProvider: any =
+    injectedProviders.find((provider) => {
+      return provider.isMetaMask ?? false
+    }) ?? (await detectEthereumProvider())
+
+  if (!injectedProvider) {
+    window.open(GET_METAMASK_LINK)
```
Comment (rzadp) [inline packages/core/src/providers/network/connectors/implementations/metamask.ts:21]: consider react-native

**Review by rzadp — APPROVED** [review]

**Pull request merged**